redhat-6.2-with-source/redhat-6.2/en_6.2/misc/src/anaconda/isys/modutils/man/lsmod.man
.\" Copyright (c) 1996 Free Software Foundation, Inc.
.\" This program is distributed according to the Gnu General Public License.
.\" See the file COPYING in the kernel source directory
.\" $Id: lsmod.1,v 1.1.1.1 1998/01/06 20:51:07 ewt Exp $
.\"
.TH LSMOD 1 "26 Dec 1996" Linux "Linux Module Support"
.SH NAME
lsmod \- list loaded modules.
.SH SYNOPSIS
.B lsmod
.SH DESCRIPTION
.B Lsmod
shows information about all loaded modules.
.PP
The format is name, size, use count, list of referring modules.  The
information displayed is identical to that available from /proc/modules.
.SH SEE ALSO
insmod(1), modprobe(1), depmod(1), rmmod(1), ksyms(1), modules(2)
.SH HISTORY
Module support was first concieved by Anonymous
.br
Initial Linux version by Bas Laarhoven <bas@vimec.nl>
.br
Version 0.99.14 by Jon Tombs <jon@gtex02.us.es>
.br
Extended by Bjorn Ekwall <bj0rn@blox.se>
.br
Updated for 2.1.17 by Richard Henderson <rth@tamu.edu>


modutils/ChangeLog
Thu Sep 11 08:43:03 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* Release version 2.1.55.

Wed Sep 10 16:34:41 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* genksyms/genksyms.c (crc_prefix): New variable.
	(export_symbol): Use it.
	(main): New option -p sets it.
	* man/genksyms.8: Update documentation.
	Idea from Jacques Gelinas <jack@solucorp.qc.ca>.

Wed Sep 10 16:09:37 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* insmod/Makefile: Allow insmod & rmmod to run as a combined binary.
	* insmod/main.c: New file.
	* Makeconfig (COMBINE_INSMOD_RMMOD): New define.
	Requested by Erik Troan <ewt@redhat.com>.

Wed Sep 10 15:41:16 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* depmod/depmod.c (load_obj_file): Print error if fopen fails.
	(main): Continue processing on errors, so that we see all errors
	all at once.
	Reported by Gunther Mayer <gunther.mayer@braunschweig.netsurf.de>.

Tue Aug 26 11:14:36 -0400 1997  Telly Mavroidis  <mavroidi@ctc-fund.com>

	* util/sys_qm.c [__sparc__]: Fix typo.

Tue Jul 29 23:02:04 PDT 1997  Richard Henderson  <richard@gnu.ai.mit.edu>

	* include/module.h: Rename sizeof_type to tgt_sizeof_type.  Kill
	the struct duplication by defining a tgt_long.
	* insmod/insmod.c: Propogate sizeof_type change.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	* insmod/insmod.c (search_module_path): Add ipv6.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	Sparc64 support:
	* depmod/Makefile (DEFS): Pass in ARCH as well.
	* insmod/Makefile: Likewise.
	* obj/Makefile: Likewise.

	* insmod/insmod.c: Don't use native sizeof(type), use constants
	sizeof_type that correspond to the target.
	* include/module.h: Detect and cope with sparc64 target and
	sparc32 host.

	* include/obj.h: libc5 doesn't declare ELF64_ST_INFO either.

	* include/elf_sparc64.h: New file.
	* obj/obj_sparc64.c: New file.

Sat Jul 19 23:56:02 +0200 1997  Florian La Roche  <florian@knorke.saar.de>

	* depmod/misc.c (resolve_string): Kill assignment warning.
	* genksyms/genksyms.h: Extern outfile and debugfile properly.
	* man/modprobe.1: New file.

Wed Apr 16 15:27:17 +1000 1997  Paul Mackerras  <paulus@cs.anu.edu.au>

	* include/elf_ppc.h, obj/obj_ppc.c: New files.

Mon Apr 14 20:33:56 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (print_deps_file): Fix initial array
	allocation size.

Sat Apr 12 19:23:25 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* obj/obj_reloc.c (obj_relocate): If AXP_BROKEN_GAS, work
	around the old gas local symbol LITERAL reloc bug.

Mon Mar 31 23:03:12 EST 1997  David S. Miller  <davem@jenolan.rutgers.edu>

	* genksyms/lex.l (yylex) [case ST_TABLE_5]: Missing break.

Tue Mar 25 21:41:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (old_read_kernel_syms): Get symbols via
	get_kernel_syms syscall rather than /proc/ksyms.

	* depmod/modprobe.c (is_removable): Final arg to query_module
	is a size_t, not an int.

Sat Mar 15 19:34:07 PST 1997  H.J. Lu  <hjl@gnu.ai.mit.edu>

	* depmod/conf_file.c (release_all_sets): Set mod_set to NULL.
	(read_config_file): only call release_all_sets () once.
	call resolve_string () to expand string.

	* depmod/misc.c (resolve_string): New function that resolves
	string with `foo`.
	(what_command): New function that returns the known commands.
	(expand_command): New funtion that expands pattern or
	command. Only `uname -r` is supported this time.

	* depmod/misc.h: New prototype for resolve_string ().

	* depmod/modprobe.c (read_deps_file): call resolve_string () to
	expand string.

	* man/depmod.1: Fix typos. The syntax should be "path[TAG]=PATHNAME"
	means TAG is appended to PATHNAME.

Thu Mar 13 13:38:12 CST 1997  Richard Henderson  <rth@tamu.edu>

	* insmod/insmod.c (ncv_strcmp): New function that combines old
	m_strcmp and k_strcmp -- it's not a simple matter to always have
	the versioned symbol as the first or second argument.
	(main): Install ncv_strcmp instead.
	Reported by David Bourgin <dbourgin@wsc.com>.

	* insmod/insmod.c (get_kernel_version): Take an extra argument to
	return the full uname string.
	(old_get_module_version): Likewise.
	(new_get_module_version): Likewise.
	(main): Record and compare the full uname string, to handle
	suffixes like "-ISS" and "-SMP".

Sun Mar  9 23:52:18 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Only use the file component of the input
	path when generating the output filename.
	Reported by Markus Dickebohm <m.dickebohm@uni-koeln.de>.

	* depmod/Makefile (clean): Remove modprobe and depmod.

Sun Mar  9 13:55:27 MET 1997  Eddie C. Dost  <ecd@skynet.be>

	* obj/obj_sparc.c (arch_apply_relocation): Handle lots
	more relocation types.

Tue Mar  4 08:12:58 CST 1997  Richard Henderson  <rth@tamu.edu>

	* Merge new depmod from Marcin Dalecki
	<dalecki@sub994.sub.uni-goettingen.de>.

	* insmod/insmod.c (get_kernel_version): Don't assert that the last
	number of the version be followed by a NUL.
	(old_get_module_version, new_get_module_version): Likewise.
	Reported by Keith Owens <kaos@ocs.com.au>

	* insmod/insmod.c: Merge character and string array patches
	from Jean Tourrilhes <jt@hplb.hpl.hp.com>.

	* kerneld/kerneld.c (kerneld_error): Add missing va_end; use vsnprintf.

Sun Feb  2 13:21:39 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Update documentation on 2.0.x symbol table
	processing.  Generate phrases for asm blocks as well, as it is more
	efficient to do it in the lexer.
	* genksyms/parse.y (opt_asm_phrase, asm_declaration): Recognise 
	the ASM_PHRASE token rather than trying to work it out ourselves.

Sun Feb  2 07:44:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* man/kerneld.8: Format for section 8.
	* man/genksyms.8: Renamed from genksyms.1 to match how it formats.
	Reported by Martin von Loewis <martin@mira.isdn.cs.tu-berlin.de>.

Tue Jan 28 21:10:07 MET 1997  Janos Farkas  <chexum@shadow.banki.hu>


modutils/insmod/lsmod.c
/* List the currently loaded modules.
   Copyright 1996, 1997 Linux International.

   New implementation contributed by Richard Henderson <rth@tamu.edu>
   Based on original work by Bjorn Eckwall <bj0rn@blox.se>

   This file is part of the Linux modutils.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2 of the License, or (at your
   option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ident "$Id: lsmod.c,v 1.1.1.1 1998/01/06 20:51:07 ewt Exp $"

#include <sys/types.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>

#include "module.h"
#include "util.h"

#include "logger.h"

/*======================================================================*/

/* If we don't have query_module ... */

static int
old_lsmod(void)
{
  int fd, len, bufsize;
  char *buffer;

  /* We've got no other option but to read /proc.  */

  if ((fd = open("/proc/modules", O_RDONLY)) < 0)
    {
      error("/proc/modules: %m");
      return 1;
    }

  buffer = xmalloc(bufsize = 8*1024);
retry_read:
  len = read(fd, buffer, bufsize);
  if (len < 0)
    {
      error("/proc/modules: %m");
      return 1;
    }
  else if (len == sizeof(buffer))
    {
      lseek(fd, 0, SEEK_SET);
      buffer = xrealloc(buffer, bufsize *= 2);
      goto retry_read;
    }

  close(fd);

  /* Write it out.  */

  puts("Module         Pages    Used by");
  write(1, buffer, len);

  return 0;
}

/* If we do have query_module ... */

static int
new_lsmod(void)
{
  char *module_names, *m, *refs;
  size_t bufsize, ret, nmod, i;

  /* A header.  */

  puts("Module                  Size  Used by");

  /* Fetch the list of modules.  */

  module_names = xmalloc(bufsize = 1024);
retry_mod_load:
  if (query_module(NULL, QM_MODULES, module_names, bufsize, &ret))
    {
      if (errno == ENOSPC)
	{
	  module_names = xrealloc(module_names, bufsize = ret);
	  goto retry_mod_load;
	}
      error("QM_MODULES: %m");
      return 1;
    }
  nmod = ret;

  refs = xmalloc(bufsize = 1024);
  for (i = 0, m = module_names; i < nmod; ++i, m += strlen(m)+1)
    {
      struct new_module_info info;
      size_t j;
      char *r;

      if (query_module(m, QM_INFO, &info, sizeof(info), &ret))
	{
	  error("QM_INFO: %m");
	  return 1;
	}

      printf("%-20s%8lu%4ld ", m, info.size, info.usecount);

      if (info.flags & NEW_MOD_DELETED)
	fputs(" (deleted)", stdout);
      else if (!(info.flags & NEW_MOD_RUNNING))
	fputs(" (uninitialized)", stdout);
      else 
	{
	  if (info.flags & NEW_MOD_AUTOCLEAN)
	    fputs(" (autoclean)", stdout);
	  if (!(info.flags & NEW_MOD_USED_ONCE))
	    fputs(" (unused)", stdout);
	}

    retry_ref_load:
      if (query_module(m, QM_REFS, refs, bufsize, &ret))
	{
	  if (errno == ENOSPC)
	    {
	      refs = xrealloc(refs, bufsize = ret);
	      goto retry_ref_load;
	    }
	  error("QM_REFS: %m");
	  return 1;
	}

      if (ret > 0)
	{
	  putchar(' ');
	  putchar('[');
	  j = 0, r = refs;
	  fputs(r, stdout);
	  while (r += strlen(r)+1, ++j < ret)
	    {
	      putchar(' ');
	      fputs(r, stdout);
	    }
	  putchar(']');
	}

      putchar('\n');
    }

  return 0;
}

int
main(int argc, char **argv)
{
  error_file = "lsmod";
  if (query_module(NULL, 0, NULL, 0, NULL) == 0)
    return new_lsmod();
  else
    return old_lsmod();
}
